parsing/parser.mly

```diff
@@ -469,27 +470,30 @@ type let_bindings =
     lbs_extension: string Asttypes.loc option;
     lbs_loc: Location.t }
 
-let mklb first ~loc (p, e) attrs =
+let mklb first ~loc (p, e, is_pun) attrs =
   {
     lb_pattern = p;
     lb_expression = e;
+    lb_is_pun = is_pun;
     lb_attributes = attrs;
     lb_docs = symbol_docs_lazy loc;
     lb_text = (if first then empty_text_lazy
                else symbol_text_lazy (fst loc));
     lb_loc = make_loc loc;
   }
 
+let addlb lbs lb =
+  if lb.lb_is_pun && lbs.lbs_extension = None then syntax_error ();
+  { lbs with lbs_bindings = lb :: lbs.lbs_bindings }
```

Comment:
Why are we checking validity at this point, rather than during the construction of `lb`? I suppose that there is a non-trivial reason which would deserve a comment.

Reply:
Nope, but there is a trivial reason: `lb` has `lb_is_pun` but `lbs` has `lbs_extension`, so you can only check this when combining `lb` with `lbs`.

This is different from binding operators: they repeat the operator on each `and` while extension nodes do not:
```ocaml
let* x = 1 and* y = 2 and* z = 3 in ...
let%foo x = 1 and y = 2 and z = 3 in ...
```

Reply:
Ah, thanks.